src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2147 @@
+     and_exprt(not_exprt(equal_exprt(str[idx], space_char)),
+               not_exprt(space_before)));
+  axioms.push_back(a9);
+  return res;
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_to_lower_case
+(const function_application_exprt &expr)
+{
+  string_exprt str=add_axioms_for_string_expr(args(expr, 1)[0]);
+  string_exprt res(get_char_type());
+  exprt char_a=constant_char('a');
+  exprt char_A=constant_char('A');
+  exprt char_z=constant_char('z');
+  exprt char_Z=constant_char('Z');
+
+  // We add axioms:
+  // a1 : |res| = |str|
+  // a2 : forall idx<str.length, 'A'<=str[idx]<='Z' => res[idx]=str[idx]+'a'-'A'
+  // a3 : forall idx<str.length, !('a'<=str[idx]<='z') => res[idx]=str[idx]
+  // forall idx<str.length,
+  // this[idx]='A'<=str[idx]<='Z' ? str[idx]+'a'-'A' : str[idx]
+
+  exprt a1=res.same_length(str);
+  axioms.push_back(a1);
+
+  symbol_exprt idx=fresh_univ_index("QA_lower_case");
+  exprt is_upper_case=and_exprt(binary_relation_exprt(char_A, ID_le, str[idx]),
+                                binary_relation_exprt(str[idx], ID_le, char_Z));
+  minus_exprt diff(char_a, char_A);
+  equal_exprt convert(res[idx], plus_exprt(str[idx], diff));
+  string_constraintt a2(idx, res.length(), is_upper_case, convert);
+  axioms.push_back(a2);
+
+  equal_exprt eq(res[idx], str[idx]);
+  string_constraintt a3(idx, res.length(), not_exprt(is_upper_case), eq);
+  axioms.push_back(a3);
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_to_upper_case
+(const function_application_exprt &expr)
+{
+  string_exprt str=add_axioms_for_string_expr(args(expr, 1)[0]);
+  string_exprt res(get_char_type());
+  exprt char_a=constant_char('a');
+  exprt char_A=constant_char('A');
+  exprt char_z=constant_char('z');
+  exprt char_Z=constant_char('Z');
+
+  // We add axioms:
+  // a1 : |res| = |str|
+  // a2 : forall idx<str.length, 'a'<=str[idx]<='z' => res[idx]=str[idx]+'A'-'a'
+  // a3 : forall idx<str.length, !('a'<=str[idx]<='z') => res[idx]=str[idx]
+  exprt a1=res.same_length(str);
+  axioms.push_back(a1);
+
+  symbol_exprt idx=fresh_univ_index("QA_upper_case");
+  exprt is_lower_case=and_exprt(binary_relation_exprt(char_a, ID_le, str[idx]),
+                                binary_relation_exprt(str[idx], ID_le, char_z));
+  minus_exprt diff(char_A, char_a);
+  equal_exprt convert(res[idx], plus_exprt(str[idx], diff));
+  string_constraintt a2(idx, res.length(), is_lower_case, convert);
+  axioms.push_back(a2);
+
+  equal_exprt eq(res[idx], str[idx]);
+  string_constraintt a3(idx, res.length(), not_exprt(is_lower_case), eq);
+  axioms.push_back(a3);
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_from_int
+(const function_application_exprt &expr)
+{ return add_axioms_from_int(args(expr, 1)[0], 10); }
+
+string_exprt string_constraint_generatort::add_axioms_from_long
+(const function_application_exprt &expr)
+{ return add_axioms_from_int(args(expr, 1)[0], 30); }
```

Comment:
What is the meaning of this magic number "30" ? Shouldn't you compute the maximum length of a long?